File: old/actions/sysctl_override.py
"""
Systemd sysctl override action
"""
import logging
from cli import FileOps
from .base import Action
logger = logging.getLogger(__name__)

class SysctlOverrideAction(Action):
    """Action to configure systemd sysctl override"""
    description = "systemd sysctl override"

    def execute(self) -> bool:
        """Configure systemd sysctl override"""
        if not self.pct_service or not self.container_id:
            logger.error("PCT service or container ID not available")
            return False
        mkdir_cmd = FileOps().mkdir("/etc/systemd/system/systemd-sysctl.service.d", parents=True)
        output, exit_code = self.pct_service.execute(self.container_id, mkdir_cmd)
        if exit_code is not None and exit_code != 0:
            logger.error("create sysctl override directory failed with exit code %s", exit_code)
            if output:
                logger.error("create sysctl override directory output: %s", output.splitlines()[-1])
            return False
        override_cmd = FileOps().write(
            "/etc/systemd/system/systemd-sysctl.service.d/override.conf",
            "[Service]\nImportCredential=\n",
        )
        output, exit_code = self.pct_service.execute(self.container_id, override_cmd)
        if exit_code is not None and exit_code != 0:
            logger.error("write sysctl override failed with exit code %s", exit_code)
            if output:
                logger.error("write sysctl override output: %s", output.splitlines()[-1])
            return False
        reload_cmd = (
            "systemctl daemon-reload && "
            "systemctl stop systemd-sysctl.service 2>/dev/null || true && "
            "systemctl start systemd-sysctl.service 2>/dev/null || true"
        )
        output, exit_code = self.pct_service.execute(self.container_id, reload_cmd)
        if exit_code is not None and exit_code != 0:
            logger.error("reload systemd-sysctl failed with exit code %s", exit_code)
            if output:
                logger.error("reload systemd-sysctl output: %s", output.splitlines()[-1])
            return False
        return True

File: src/actions/sysctl_override.go
package actions

import (
	"strconv"
	"strings"
	"enva/cli"
	"enva/libs"
	"enva/services"
)

// SysctlOverrideAction configures systemd sysctl override
type SysctlOverrideAction struct {
	*BaseAction
}

func NewSysctlOverrideAction(sshService *services.SSHService, aptService *services.APTService, pctService *services.PCTService, containerID *string, cfg *libs.LabConfig, containerCfg *libs.ContainerConfig) Action {
	return &SysctlOverrideAction{
		BaseAction: &BaseAction{
			SSHService:   sshService,
			APTService:   aptService,
			PCTService:   pctService,
			ContainerID: containerID,
			Cfg:          cfg,
			ContainerCfg: containerCfg,
		},
	}
}

func (a *SysctlOverrideAction) Description() string {
	return "systemd sysctl override"
}

func (a *SysctlOverrideAction) Execute() bool {
	if a.PCTService == nil || a.ContainerID == nil {
		libs.GetLogger("sysctl_override").Printf("PCT service or container ID not available")
		return false
	}
	containerIDInt, err := strconv.Atoi(*a.ContainerID)
	if err != nil {
		libs.GetLogger("sysctl_override").Printf("Invalid container ID: %s", *a.ContainerID)
		return false
	}
	mkdirCmd := cli.NewFileOps().Mkdir("/etc/systemd/system/systemd-sysctl.service.d", true)
	output, exitCode := a.PCTService.Execute(containerIDInt, mkdirCmd, nil)
	if exitCode != nil && *exitCode != 0 {
		libs.GetLogger("sysctl_override").Printf("create sysctl override directory failed with exit code %d", *exitCode)
		if output != "" {
			lines := strings.Split(output, "\n")
			if len(lines) > 0 {
				libs.GetLogger("sysctl_override").Printf("create sysctl override directory output: %s", lines[len(lines)-1])
			}
		}
		return false
	}
	overrideCmd := cli.NewFileOps().Write("/etc/systemd/system/systemd-sysctl.service.d/override.conf", "[Service]\nImportCredential=\n")
	output, exitCode = a.PCTService.Execute(containerIDInt, overrideCmd, nil)
	if exitCode != nil && *exitCode != 0 {
		libs.GetLogger("sysctl_override").Printf("write sysctl override failed with exit code %d", *exitCode)
		if output != "" {
			lines := strings.Split(output, "\n")
			if len(lines) > 0 {
				libs.GetLogger("sysctl_override").Printf("write sysctl override output: %s", lines[len(lines)-1])
			}
		}
		return false
	}
	reloadCmd := "systemctl daemon-reload && systemctl stop systemd-sysctl.service || true && systemctl start systemd-sysctl.service || true"
	output, exitCode = a.PCTService.Execute(containerIDInt, reloadCmd, nil)
	if exitCode != nil && *exitCode != 0 {
		libs.GetLogger("sysctl_override").Printf("reload systemd-sysctl failed with exit code %d", *exitCode)
		if output != "" {
			lines := strings.Split(output, "\n")
			if len(lines) > 0 {
				libs.GetLogger("sysctl_override").Printf("reload systemd-sysctl output: %s", lines[len(lines)-1])
			}
		}
		return false
	}
	return true
}

